include/deal.II/matrix_free/face_setup_internal.h

```diff
@@ -0,0 +1,944 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2018 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii__face_setup_internal_h
+#define dealii__face_setup_internal_h
+
+
+#include <deal.II/base/utilities.h>
+#include <deal.II/base/memory_consumption.h>
+#include <deal.II/grid/tria.h>
+#include <deal.II/grid/tria_accessor.h>
+#include <deal.II/matrix_free/task_info.h>
+#include <deal.II/matrix_free/face_info.h>
+
+#include <fstream>
+
+
+DEAL_II_NAMESPACE_OPEN
+
+
+namespace internal
+{
+  namespace MatrixFreeFunctions
+  {
+    /**
+     * A struct that is used to represent a collection of faces of a process
+     * with one of its neighbor within the setup done in struct FaceInfo.
+     *
+     * @author Katharina Kormann, Martin Kronbichler, 2018
+     */
+    struct FaceIdentifier
+    {
+      FaceIdentifier()
+        :
+        n_hanging_faces_smaller_subdomain (0),
+        n_hanging_faces_larger_subdomain  (0)
+      {}
+
+      std::vector<std::pair<CellId,CellId> > shared_faces;
+      unsigned int n_hanging_faces_smaller_subdomain;
+      unsigned int n_hanging_faces_larger_subdomain;
+    };
+
+
+
+    /**
+     * A struct that extracts the faces relevant to a given set of cells,
+     * including the assignment of which of the two neighboring processors at
+     * a subdomain boundary with MPI should do the integration (from both
+     * sides). This data structure is used for the setup of the connectivity
+     * between faces and cells and for identification of the dof indices to be
+     * used for face integrals.
+     *
+     * @author Katharina Kormann, Martin Kronbichler, 2018
+     */
+    template <int dim>
+    struct FaceSetup
+    {
+      FaceSetup()
+        :
+        use_active_cells (true)
+      {}
+
+      /**
+       * Perform the initial detection of faces before reading the indices on
+       * the cells. This does not add the faces yet but only decides on
+       * whether some of the faces should be considered for processing
+       * locally.
+       */
+      template <typename MFAddData>
+      void initialize(const dealii::Triangulation<dim> &triangulation,
+                      const MFAddData                  &additional_data,
+                      std::vector<std::pair<unsigned int,unsigned int> > &cell_levels)
+      {
```

Comment:
Could we separate the implementation of the functions from the definition (just move it at the end of the file). I am trying to understand the interface of the structure and which data is stored and I spend my time scrolling up and down.